streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -241,16 +239,29 @@ private static void assignStatelessActiveTasks(final TreeMap<UUID, ClientState>
         final Map<TaskId, SortedSet<UUID>> taskToCaughtUpClients = new HashMap<>();
 
         for (final TaskId task : statefulTasks) {
-
+            final TreeSet<UUID> caughtUpClients = new TreeSet<>();
             for (final Map.Entry<UUID, ClientState> clientEntry : clientStates.entrySet()) {
                 final UUID client = clientEntry.getKey();
                 final long taskLag = clientEntry.getValue().lagFor(task);
-                if (taskLag == Task.LATEST_OFFSET || taskLag <= acceptableRecoveryLag) {
-                    taskToCaughtUpClients.computeIfAbsent(task, ignored -> new TreeSet<>()).add(client);
+                if (active(taskLag) || unbounded(acceptableRecoveryLag) || acceptable(acceptableRecoveryLag, taskLag)) {
```

Comment:
I realized that our condition was actually wrong here. In addition to all the zero-or-greater lags, there are two negative lags, one meaning "unknown" (-1), and one meaning "latest" (-2). When we said `taskLag <= acceptableRecoveryLag`, it was unintentionally encompassing the sentinel values as well. Even if we want a sentinel to be considered "caught up" (as with "Latest"), we should do so explicitly, not by mathematical coincidence.

I also added a special case when acceptableRecoveryLag is set to MAX_VALUE to indicate that all tasks, regardless of their lag (even if it's a sentinel), are to be considered caught-up.

I also found the boolean expression with all the conditionals a little hard to read, so I pulled out some semantic methods.

Reply:
Nice catch indeed! I agree with @ableegoldman about the renaming. I am in favour of `activeRunning` or `activeAndRunning`.